src/main/java/com/vmware/vipclient/i18n/messages/service/LocaleService.java

```diff
@@ -31,119 +34,166 @@ public LocaleService(LocaleDTO dto) {
-        if (regionMap != null) {
-            return regionMap;
+        LocaleCacheItem cacheItem = getRegionsByLocale(locale);
+        if (!cacheItem.getCachedData().isEmpty()) {
+            return cacheItem.getCachedData();
         }
         Iterator<Locale> fallbackLocalesIter = LocaleUtility.getFallbackLocales().iterator();
         while (fallbackLocalesIter.hasNext()) {
             String fallbackLocale = fallbackLocalesIter.next().toLanguageTag();
             if(fallbackLocale.equalsIgnoreCase(locale))
                 continue;
             logger.info("Can't find regions for locale [{}], look for fallback locale [{}] regions as fallback!", locale, fallbackLocale);
-            regionMap = getRegionsByLocale(fallbackLocale);
-            if (regionMap != null) {
-                new FormattingCacheService().addRegions(locale, regionMap);
-                logger.debug("Fallback locale [{}] regions is cached for locale [{}]!\n\n", fallbackLocale, locale);
+            cacheItem = getRegionsByLocale(fallbackLocale);
+            if (!cacheItem.getCachedData().isEmpty()) {
                 break;
             }
         }
-        return regionMap;
+        return cacheItem.getCachedData();
     }
 
-    public Map<String, String> getRegionsByLocale(String locale){
+    public LocaleCacheItem getRegionsByLocale(String locale){
         if(locale != null && !locale.isEmpty())
             locale = locale.replace("_", "-").toLowerCase();
-        Map<String, String> regionMap = null;
+        LocaleCacheItem cacheItem = null;
         logger.debug("Look for region list from cache for locale [{}]", locale);
         FormattingCacheService formattingCacheService = new FormattingCacheService();
-        regionMap = formattingCacheService.getRegions(locale);
-        if (regionMap != null) {
+        cacheItem = formattingCacheService.getRegions(locale);
+        if (cacheItem != null) {
+            if (cacheItem.isExpired()) { // cacheItem has expired
+                // Update the cache in a separate thread
+                populateRegionsCache(locale, cacheItem);
+            }
             logger.debug("Find regions from cache for locale [{}]!", locale);
-            return regionMap;
-        }
-        regionMap = getRegionsFromDS(locale, VIPCfg.getInstance().getMsgOriginsQueue().listIterator());
-        if (regionMap != null) {
-            logger.debug("Find the regions for locale [{}].\n", locale);
-            formattingCacheService.addRegions(locale, regionMap);
-            logger.debug("Regions is cached for locale [{}]!\n\n", locale);
-            return regionMap;
+            return cacheItem;
         }
-        return regionMap;
+        cacheItem = new LocaleCacheItem();
+        getRegionsFromDS(locale, cacheItem, VIPCfg.getInstance().getMsgOriginsQueue().listIterator());
+        formattingCacheService.addRegions(locale, cacheItem);
+        logger.debug("Regions is cached for locale [{}]!\n\n", locale);
```

Comment:
This is for handling non-supported locales, just want to add an empty fetch result to cache, it's the same as you new an empty MessageCacheItem and add it to cache.

Reply:
If you noticed in my most recent PR, we have changed the design of caching empty cacheItem objects for non-supported locales. This is because this design won't help in the case of the very first request (it will still be slow). So for L3, we just use the list of supported locales to figure out the next steps.. and in the case of non-supported locale, that would be to return fallback locale messages.

For L2, can we do the same in next PR? For now, add the if statement.